Fix #17308: Fixed version error while editing question.

kshitij01042002 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #17308.
2. This PR does the following: Added schema for question version. Also validated versions of the questions while saving the question.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->
Can edit and save question now.


[Screencast from 19-02-23 12:42:24 PM IST.webm](https://user-images.githubusercontent.com/91470808/219936864-5556581e-42be-428a-873e-c934d3e792b0.webm)

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

kshitij01042002 committed:
failing tests

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
restored changed file

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Fixed version error

kshitij01042002 committed:
Fixed version error

kshitij01042002 committed:
Fixed mypy errors

kshitij01042002 committed:
Fixed mypy errors

kshitij01042002 committed:
Fixed mypy errors

kshitij01042002 commented:
@aks681 since the files changed are not of frontend I am unsure about which changelog label to add, need Guidance.
Is changelog label of Learners/Creators page correct?
Thanks.

kshitij01042002 committed:
fixed backend tests

kshitij01042002 committed:
added full coverage to backend tests

kshitij01042002 committed:
fixed lint checks

kshitij01042002 committed:
fix lint errors

kshitij01042002 committed:
fixed backend coverage

kshitij01042002 committed:
Merge branch 'develop' into edit_question

kshitij01042002 commented:
@aks681 the PR is getting stale, can you PTAL?

Thanks!

aks681 commented on core/controllers/question_editor.py:
```diff
@@ -329,12 +340,22 @@ def put(self, question_id: str) -> None:
+                    'Cannot create a new fully specified question'
+                )
 
         question_services.update_question(
             self.user_id, question_id, change_list,
```
Can't the above check be done inside the update function, when the question is fetched? Since, the question is being fetched in the controller just for this check.

kshitij01042002 replied:
If we do the above check in the update function, I am not able to mock the test where version of the incoming question and current question differ to write the test cases.
I believe the versioning model for question is different for that of other models of the similar type.

aks681 replied:
Ok, so normally testing shouldn't be the main reason to choose one way of implementing things over another. 
Also, in this case, can't you pass the 'version' from the payload as an additional argument to the update question function, and do the comparison in there? This would help in mocking it as well.

kshitij01042002 replied:
Soo should I remove validation from the controller? Cause it is done in other versioning models like topic/story.

kshitij01042002 committed:
Temporary changes

kshitij01042002 committed:
Merge branch 'develop' into edit_question

kshitij01042002 committed:
Merge branch 'edit_question' of https://github.com/kshitij01042002/oppia into edit_question

kshitij01042002 committed:
new changes

kshitij01042002 committed:
removed validation from controller

kshitij01042002 committed:
fixed lint checks

kshitij01042002 committed:
fixed lint checks

kshitij01042002 committed:
fixed lint checks

kshitij01042002 committed:
fixed mypy checks

kshitij01042002 commented:
@aks681 I have removed validation from the controller and added it in the update function as you suggested. Can you PTAL?

Thanks!


Proof that validation works:-
[Screencast from 01-03-23 09:56:21 AM IST.webm](https://user-images.githubusercontent.com/91470808/222050760-711966dd-e747-4bee-9fd1-456ca43a0632.webm)

kshitij01042002 committed:
Merge branch 'develop' into edit_question

SubhashThenua commented:
Hi @kshitij01042002 
FYI -- If the reviewer does not response in 24 hours
 -- If you need review from Reviewers
Please tag the reviewer { {@Reviewer name }} PTAL.
Thanks!

kshitij01042002 commented:
@aks681 PTAL.

Thanks

kshitij01042002 commented:
Thanks @SubhashThenua for your suggestion but I had already mailed akshay regarding this.

kshitij01042002 committed:
Merge branch 'develop' into edit_question

aks681 commented on core/domain/question_services.py:
```diff
@@ -756,14 +759,28 @@ def update_question(
         raise ValueError(
             'Expected a commit message, received none.')
     updated_question = apply_change_list(question_id, change_list)
+    if (version is not None) and (version < updated_question.version):
```
Just to confirm, if the current version of question is 4 and it got updated to 5. Won't these values be 4 and 5, but the condition shouldn't throw an error? or is the 'version' value in payload the incremented version?

kshitij01042002 replied:
Actually Here the `updated_question.version` still have the version of the current question  and `version` have the version coming from the payload. `apply_change_list` functions changes other values of the question but it's version is updated when a _save_question is called

aks681 replied:
Sorry for the delay.
Ah ok, thanks for confirming, lgtm!

kshitij01042002 commented:
@aks681 can you PTAL?

Thanks.

kshitij01042002 committed:
Merge branch 'develop' into edit_question

kshitij01042002 committed:
Merge branch 'develop' into edit_question

kshitij01042002 commented:
Hii @aks681 just a friendly reminder to take a look at the PR as Build of the PR is getting STALE.

Thanks.
